nemo/collections/asr/metrics/wer.py

```diff
@@ -52,88 +49,105 @@ def word_error_rate(hypotheses: List[str], references: List[str], use_cer=False)
-    """
-    Decodes a sequence of labels to words
-    """
-    blank_id = len(labels)
-    hypotheses = []
-    labels_map = dict([(i, labels[i]) for i in range(len(labels))])
-    prediction_cpu_tensor = tensor.long().cpu()
-    # iterate over batch
-    for ind in range(prediction_cpu_tensor.shape[0]):
-        prediction = prediction_cpu_tensor[ind].detach().numpy().tolist()
-        # CTC decoding procedure
-        decoded_prediction = []
-        previous = len(labels)  # id of a blank symbol
-        for p in prediction:
-            if (p != previous or previous == blank_id) and p != blank_id:
-                decoded_prediction.append(p)
-            previous = p
-        hypothesis = ''.join([labels_map[c] for c in decoded_prediction])
-        hypotheses.append(hypothesis)
-    return hypotheses
-
-
-def monitor_asr_train_progress(tensors: list, labels: list, eval_metric='WER', ctc_decode=True):
+class AverageTextWER(TensorMetric):
     """
-    Takes output of greedy ctc decoder and performs ctc decoding algorithm to
-    remove duplicates and special symbol. Prints sample to screen, computes
-    and logs AVG WER to console
+    This metric computes nominator and denominator for Average Text Word Error Rate.
+    When doing distributed training/evaluation the result of res=AverageTextWER(predictions, targets, target_lengths) calls
+    will be all-reduced between all workers using SUM operations.
+    Here contains two numbers res=[wer_nominator, wer_denominator]
+
+    If used with PytorchLightning LightningModule, include wer_nominator and wer_denominator inside validation_step results.
+    Then aggregate (sum) then at the end of validation epoch to correctly compute validation WER.
+
+    Example:
+        def validation_step(self, batch, batch_idx):
+            ...
+            wer_num, wer_denum = self.__wer(predictions, transcript, transcript_len)
+            return {'val_loss': loss_value, 'val_wer_num': wer_num, 'val_wer_denum': wer_denum}
+
+        def validation_epoch_end(self, outputs):
+            ...
+            wer_num = torch.stack([x['val_wer_num'] for x in outputs]).sum()
+            wer_denum = torch.stack([x['val_wer_denum'] for x in outputs]).sum()
+            tensorboard_logs = {'validation_loss': val_loss_mean, 'validation_avg_wer': wer_num / wer_denum}
+            return {'val_loss': val_loss_mean, 'log': tensorboard_logs}
+
     Args:
-      tensors: A list of 3 tensors (predictions, targets, target_lengths)
-      labels: A list of labels
-      eval_metric: An optional string from 'WER', 'CER'. Defaults to 'WER'.
+        vocabulary:
+        batch_dim_index:
+        use_cer:
+        ctc_decode:
+
     Returns:
-      batch wer, hypothesis and reference from the first batch element
+        res: a torch.Tensor object with two elements: [wer_nominator, wer_denominator]. To correctly compute average
+        text word error rate, compute wer=wer_nominator/wer_denominator
```

Comment:
Nitpick: `s/nominator/numerator/` and `s/wer_num, wer_denum/wer_num, wer_denom/`